api/src/services/security-service.ts

```diff
@@ -76,4 +84,86 @@ export class SecurityService extends DBService {
       )
     );
   }
+
+  /**
+   * Returns the signed URL of a document for which a user has permissions.
+   *
+   * Rules:
+   * non-admin user cannot access the document when:
+   ** - it is pending review, OR
+   ** - user hasn't been granted an exception to every security rule
+   *
+   * non-admin user can access the document when:
+   * - document is not secured
+   * - user has the correct exceptions
+   *
+   *
+   * @param {number} artifactId
+   * @return {*}  {Promise<any>}
+   * @memberof SecurityService
+   */
+  async getSecuredArtifactBasedOnRulesAndPermissions(artifactId: number): Promise<any> {
+    const isSystemUserAdmin = await this.userService.isSystemUserAdmin();
+
+    const userId = this.connection.systemUserId();
+
+    const isDocumentPendingReview = (await this.artifactService.getArtifactById(artifactId)).security_review_timestamp
+      ? false
+      : true;
+
+    if (!isSystemUserAdmin && isDocumentPendingReview) {
+      throw new HTTP403('You do not have access to this document - it is pending review');
+    }
+
+    const documentSecurityRules = await this.getDocumentPersecutionAndHarmRules(artifactId);
+
+    const pers_harm_exceptions = await this.getPersecutionAndHarmExceptionsByUser(userId);
+
+    const userHasExceptionsToAllRules = documentSecurityRules.every((rule) => pers_harm_exceptions.includes(rule));
+
+    if (
+      !isSystemUserAdmin &&
+      !isDocumentPendingReview &&
+      documentSecurityRules.length > 0 &&
+      !userHasExceptionsToAllRules
+    ) {
+      throw new HTTP403('You do not have access to this document');
+    }
+
+    // access is granted because
+    // 1) admin
+    // 2) document is unsecured (not pending review, and has no security rules)
+    // 3) non-admin has exceptions all security rules
+    const response = await this.artifactService.getArtifactById(artifactId);
+    return await getS3SignedURL(response.key);
```

Comment:
Nit: can remove the `await`, since `getS3SignedUrl` and `await getS3SignedUrl` return the same thing

Reply:
@curtisupshall good catch, thank you